File: frontend/src/lib/assets.ts
/**
 * Asset inventory & connector API client — M3.
 *
 * Tenant-scoped throughout via the existing bearer-token session; no
 * organization_id is ever supplied client-side.
 */
import { api } from "./api";

export interface Asset {
  id: string;
  organization_id: string;
  asset_type: string;
  name: string;
  description: string;
  external_id: string;
  discovery_source: string;
  lifecycle_stage: string;
  health_status: string;
  first_observed_at: string;
  last_observed_at: string;
  relationship_count: number;
  associated_target_id: string | null;
}

export interface AssetRelationship {
  relationship_id: string;
  target_asset_id: string;
  relationship_type: string;
  label: string;
}

export interface Connector {
  id: string;
  organization_id: string;
  connector_type: string;
  name: string;
  status: string;
  last_discovery_status: string | null;
  last_discovery_completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface DiscoveryRun {
  job_id: string;
  connector_id: string;
  status: string;
  started_at: string;
  completed_at: string | null;
  assets_discovered: number;
  assets_normalized: number;
  assets_failed: number;
  error_message: string | null;
}

export async function listAssets(): Promise<Asset[]> {
  return api.get<Asset[]>("/api/v1/assets?limit=100");
}

export async function getAsset(assetId: string): Promise<Asset> {
  return api.get<Asset>(`/api/v1/assets/${assetId}`);
}

export async function getAssetRelationships(assetId: string): Promise<AssetRelationship[]> {
  return api.get<AssetRelationship[]>(`/api/v1/assets/${assetId}/relationships`);
}

export async function listConnectors(): Promise<Connector[]> {
  return api.get<Connector[]>("/api/v1/connectors");
}

export async function registerConnector(name: string, description: string): Promise<Connector> {
  return api.post<Connector>("/api/v1/connectors", { name, description });
}

export async function enableConnector(connectorId: string): Promise<Connector> {
  return api.post<Connector>(`/api/v1/connectors/${connectorId}/enable`);
}

export async function disableConnector(connectorId: string): Promise<Connector> {
  return api.post<Connector>(`/api/v1/connectors/${connectorId}/disable`);
}

export async function startDiscovery(connectorId: string): Promise<DiscoveryRun> {
  return api.post<DiscoveryRun>(`/api/v1/connectors/${connectorId}/discover`);
}

export async function listDiscoveryRuns(connectorId: string): Promise<DiscoveryRun[]> {
  return api.get<DiscoveryRun[]>(`/api/v1/connectors/${connectorId}/discovery-runs`);
}

File: backend/src/redforge/api/v1/assets.py
"""Asset inventory REST API — M3.

Tenant-scoped throughout: `organization_id` is ALWAYS `tenant.organization_id`
from the verified JWT (never a client-supplied value). A cross-tenant
guessed asset ID returns 404, identically to a nonexistent one — see
`TenantAssetService.get_for_org` / `SqlAlchemyAssetRepository.get_by_id_for_org`,
which filter at the query level rather than fetching-then-checking.
"""

from __future__ import annotations

import dataclasses
from typing import TYPE_CHECKING

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel

from redforge.api.dependencies import get_tenant_asset_service
from redforge.api.security import TenantContext, require_permission
from redforge.core.exceptions import NotFoundError
from redforge.domain.identity.value_objects import Permission

if TYPE_CHECKING:
    from redforge.application.inventory.tenant_asset_service import TenantAssetService

router = APIRouter(prefix="/assets", tags=["assets"])


class AssetResponse(BaseModel):
    id: str
    organization_id: str
    asset_type: str
    name: str
    description: str
    external_id: str
    discovery_source: str
    lifecycle_stage: str
    health_status: str
    first_observed_at: str
    last_observed_at: str
    relationship_count: int
    associated_target_id: str | None


class AssetRelationshipResponse(BaseModel):
    relationship_id: str
    target_asset_id: str
    relationship_type: str
    label: str


@router.get("", response_model=list[AssetResponse])
async def list_assets(
    asset_type: str | None = Query(default=None),
    lifecycle_stage: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    tenant: TenantContext = Depends(require_permission(Permission.TARGETS_READ)),
    service: TenantAssetService = Depends(get_tenant_asset_service),
) -> list[AssetResponse]:
    """List canonical assets owned by the caller's organization."""
    assets = await service.list_for_org(
        tenant.organization_id, asset_type, lifecycle_stage, limit, offset,
    )
    return [AssetResponse(**dataclasses.asdict(a)) for a in assets]


@router.get("/{asset_id}", response_model=AssetResponse)
async def get_asset(
    asset_id: str,
    tenant: TenantContext = Depends(require_permission(Permission.TARGETS_READ)),
    service: TenantAssetService = Depends(get_tenant_asset_service),
) -> AssetResponse:
    """404s identically for a nonexistent ID or one owned by another
    organization — a cross-tenant caller cannot distinguish the two."""
    try:
        asset = await service.get_for_org(asset_id, tenant.organization_id)
    except NotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=exc.message) from exc
    return AssetResponse(**dataclasses.asdict(asset))


@router.get("/{asset_id}/relationships", response_model=list[AssetRelationshipResponse])
async def get_asset_relationships(
    asset_id: str,
    tenant: TenantContext = Depends(require_permission(Permission.TARGETS_READ)),
    service: TenantAssetService = Depends(get_tenant_asset_service),
) -> list[AssetRelationshipResponse]:
    try:
        rels = await service.get_relationships_for_org(asset_id, tenant.organization_id)
    except NotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=exc.message) from exc
    return [AssetRelationshipResponse(**r) for r in rels]
